alvr/server/cpp/platform/linux/EncodePipelineNvEnc.cpp

```diff
@@ -0,0 +1,108 @@
+#include "EncodePipelineNvEnc.h"
+#include "ALVR-common/packet_types.h"
+#include "alvr_server/Settings.h"
+#include "ffmpeg_helper.h"
+#include <chrono>
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libavutil/opt.h>
+#include <libswscale/swscale.h>
+}
+
+namespace {
+
+const char *encoder(ALVR_CODEC codec) {
+    switch (codec) {
+    case ALVR_CODEC_H264:
+        return "h264_nvenc";
+    case ALVR_CODEC_H265:
+        return "hevc_nvenc";
+    }
+    throw std::runtime_error("invalid codec " + std::to_string(codec));
+}
+
+} // namespace
+alvr::EncodePipelineNvEnc::EncodePipelineNvEnc(std::vector<VkFrame> &input_frames,
+                                               VkFrameCtx &vk_frame_ctx) {
+    int err;
+    for (auto &input_frame : input_frames) {
+        vk_frames.push_back(input_frame.make_av_frame(vk_frame_ctx).release());
+    }
+
+    const auto &settings = Settings::Instance();
+
+    auto codec_id = ALVR_CODEC(settings.m_codec);
+    const char *encoder_name = encoder(codec_id);
+    AVCodec *codec = AVCODEC.avcodec_find_encoder_by_name(encoder_name);
+    if (codec == nullptr) {
+        throw std::runtime_error(std::string("Failed to find encoder ") + encoder_name);
+    }
+
+    encoder_ctx = AVCODEC.avcodec_alloc_context3(codec);
+    if (not encoder_ctx) {
+        throw std::runtime_error("failed to allocate NVEnc encoder");
+    }
+
+    switch (codec_id) {
+    case ALVR_CODEC_H264:
+        AVUTIL.av_opt_set(encoder_ctx, "preset", "llhq", 0);
+        AVUTIL.av_opt_set(encoder_ctx, "zerolatency", "1", 0);
+        break;
+    case ALVR_CODEC_H265:
+        AVUTIL.av_opt_set(encoder_ctx, "preset", "llhq", 0);
+        AVUTIL.av_opt_set(encoder_ctx, "zerolatency", "1", 0);
+        break;
+    }
+
+    /**
+     * We will recieve a frame from HW as AV_PIX_FMT_VULKAN which will converted to AV_PIX_FMT_BGRA
+     * as SW format when we get it from HW.
+     * But NVEnc support only BGR0 format and we easy can just to force it
+     * Because:
+     * AV_PIX_FMT_BGRA - 28  ///< packed BGRA 8:8:8:8, 32bpp, BGRABGRA...
+     * AV_PIX_FMT_BGR0 - 123 ///< packed BGR 8:8:8,    32bpp, BGRXBGRX...   X=unused/undefined
+     *
+     * We just to ignore the alpha channel and it's done
+     */
+    encoder_ctx->pix_fmt = AV_PIX_FMT_BGR0;
```

Comment:
Can we add a check on input frame format? It would be better to fail with an error message than having garbage video output.

Reply:
Good point.

Reply:
Added check to init class.